### ISSUE TITLE
"IDE0072: Populate switch" showing up on complete switch expression

### ISSUE THREAD

**Issue by julealgon** [dotnet/roslyn#48876]
**Version Used**: 
- Visual Studio 2019 16.7.6

**Steps to Reproduce**:

1. Create a bool switch expression with all possible return values explicitly declared as cases:

```csharp
    public class Sample
    {
        public string Method()
        {
            var boolean = true;

            return boolean switch
            {
                true => "true",
                false => "false",
            };
        }
    }
```

2. Observe IDE0072 is triggered.

![image](https://user-images.githubusercontent.com/461579/97018202-41645600-1525-11eb-90c4-ae1ea50fd5b3.png)

**Expected Behavior**:

This analyzer should not raise the suggestion/fix for switch expressions that already contain all possible values as cases.

**Actual Behavior**:

![image](https://user-images.githubusercontent.com/461579/97018292-5b9e3400-1525-11eb-91ed-577a41fce631.png)

Adding the suggestion actually raises a CA8510 warning:

![image](https://user-images.githubusercontent.com/461579/97018433-87b9b500-1525-11eb-862d-1161859c72fd.png)

**Comment by julealgon** [comment]
Related to #37639

**Comment by JeroenOortwijn** [comment]
I have a similar issue, but in this case it pops up to add cases that are already there:

**Version Used**:
Visual Studio 2019
Version 16.8.2

**Steps to Reproduce**:
1. Start with the following code:
```
public bool IsEnabled(Microsoft.Extensions.Logging.LogLevel logLevel)
{
    switch (logLevel)
    {
        case Microsoft.Extensions.Logging.LogLevel.Trace:
        case Microsoft.Extensions.Logging.LogLevel.Debug:
        case Microsoft.Extensions.Logging.LogLevel.Information:
        case Microsoft.Extensions.Logging.LogLevel.Warning:
        case Microsoft.Extensions.Logging.LogLevel.Error:
        case Microsoft.Extensions.Logging.LogLevel.Critical:
        case Microsoft.Extensions.Logging.LogLevel.None:
            return true;
        default:
            return false;
    }
}
```
2. Apply IDE0066:
```
public bool IsEnabled(Microsoft.Extensions.Logging.LogLevel logLevel)
{
    return logLevel switch
    {
        Microsoft.Extensions.Logging.LogLevel.Trace or Microsoft.Extensions.Logging.LogLevel.Debug or Microsoft.Extensions.Logging.LogLevel.Information or Microsoft.Extensions.Logging.LogLevel.Warning or Microsoft.Extensions.Logging.LogLevel.Error or Microsoft.Extensions.Logging.LogLevel.Critical or Microsoft.Extensions.Logging.LogLevel.None => true,
        _ => false,
    };
}
```
3. Observe that IDE0072 is suggested:

  ![image](https://user-images.githubusercontent.com/51476918/99811442-7b853f80-2b45-11eb-9616-d1f3bdb1dcb0.png)

4. Apply IDE0072:
```
public bool IsEnabled(Microsoft.Extensions.Logging.LogLevel logLevel)
{
    return logLevel switch
    {
        Microsoft.Extensions.Logging.LogLevel.Trace or Microsoft.Extensions.Logging.LogLevel.Debug or Microsoft.Extensions.Logging.LogLevel.Information or Microsoft.Extensions.Logging.LogLevel.Warning or Microsoft.Extensions.Logging.LogLevel.Error or Microsoft.Extensions.Logging.LogLevel.Critical or Microsoft.Extensions.Logging.LogLevel.None => true,
        Microsoft.Extensions.Logging.LogLevel.Trace => throw new System.NotImplementedException(),
        Microsoft.Extensions.Logging.LogLevel.Debug => throw new System.NotImplementedException(),
        Microsoft.Extensions.Logging.LogLevel.Information => throw new System.NotImplementedException(),
        Microsoft.Extensions.Logging.LogLevel.Warning => throw new System.NotImplementedException(),
        Microsoft.Extensions.Logging.LogLevel.Error => throw new System.NotImplementedException(),
        Microsoft.Extensions.Logging.LogLevel.Critical => throw new System.NotImplementedException(),
        Microsoft.Extensions.Logging.LogLevel.None => throw new System.NotImplementedException(),
        _ => false,
    };
}
```
5. Observe that the code is now broken:

  ![image](https://user-images.githubusercontent.com/51476918/99811834-f9494b00-2b45-11eb-8bd2-5da62efad05e.png)

**Comment by Piedone** [comment]
The rule also seems to disregard a discard arm:

```c#
    public enum MyEnum
    {
        One,
        Two,
        Three,
        Four,
    }

    // Wants to add a case for MyEnum.One.
    return variableContainingMyEnum switch
    {
        MyEnum.Four => ...,
        MyEnum.Three => ...,
        MyEnum.Two => ...,
        _ => ...,
    };
```

**Comment by CyrusNajmabadi** [comment]
> The rule also seems to disregard a discard arm:

Yes.  This is "populate switch".  The idea here is that if you've added new enum values, you can easily then add the cases for them to add impls for.  If we didn't offer this in the case of a discard, you would still have to add all the entries manually.

**Comment by Piedone** [comment]
I'd understand this for the refactor menu but I'm seeing this as a suggestion (or warning, if configured like that). I see, however, that [the rule](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0072) specified this exact scenario as indeed causing a violation. While I disagree with this showing up as a suggestion (the discard arm is there to handle the missing cases) my case is indeed not valid for this issue.

**Comment by CyrusNajmabadi** [comment]
> (or warning, if configured like that)

I mean... you configured it to do this... so it's doing this :)

The idea is to warn you such that if you add a new case to your switch, and you don't handle it, you get warned.  So it's exactly warning you for the case it was exactly designed to catch. :)

**Comment by CyrusNajmabadi** [comment]
> the discard arm is there to handle the missing cases

If this was the case... we'd never show anything.  Because you already get a warning for a non-exhaustive switch. So to make it exhaustive you have to add the 'catch all'.  And once you add the catch all, this new analyzer is there to see if you're missing an explicit cast for an actual enum value. :)

**Comment by Piedone** [comment]
Thanks, got it.

**Comment by CyrusNajmabadi** [comment]
you're welcome!

**Comment by joelhoisko** [comment]
Also having the same problem as @julealgon, a boolean switch sill asks me to populate it with a default case even though I already have cases for `true/false`.
I'm on .NET5:
```
$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.102
 Commit:    71365b4d42

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/5.0.102/

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  3.1.405 [/usr/share/dotnet/sdk]
  5.0.102 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

**Comment by julealgon** [comment]
@CyrusNajmabadi would you mind updating the 'Milestone' for the issue? It still shows "Backlog" there.

**Comment by CyrusNajmabadi** [comment]
@julealgon a bot will auto update that.
